### ISSUE TITLE
Confusing/incorrect error message with incoherent implementations and async blocks

### ISSUE THREAD

**Issue by syntacticsugarglider** [rust-lang/rust#67651]
```Rust
use std::future::Future;

pub trait Bound {}

pub struct Error;

impl Bound for Error {}

pub struct Wrap;

impl<T: Bound> From<T> for Wrap {
    fn from(input: T) -> Self {
        Wrap
    }
}

impl<T: Into<Wrap>> From<T> for Error {
    fn from(input: T) -> Self {
        Error
    }
}

fn fail() -> impl Future<Output = Result<(), Error>> {
    async move {
        let a: Result<(), Error> = Ok(());
        let a: () = a?;
        Ok(())
    }
}
```

This fails with `type annotations required: cannot resolve Error: std::convert::From<Error>` (E0283) instead of the expected `conflicting implementations of trait std::convert::From<Error> for type Error` (E0119). Changing `fail` to 
```Rust
fn fail() -> Result<(), Error> {
    let a: Result<(), Error> = Ok(());
    let a: () = a?;
    Ok(())
}
``` 
i.e. eliminating the async block leads to the expected error message being produced.

[playground demonstration](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d5fe05a99c552d394a8a829f66cd308e)

**Comment by tmandry** [comment]
@Centril said they'd look into minimizing this further.

**Comment by tmandry** [comment]
Also assigning @nikomatsakis to look into this along with https://github.com/rust-lang/rust/issues/66312.

**Comment by Centril** [comment]
Reduced:
```rust
trait Bound {}

struct Error;

impl Bound for Error {}

impl<T> From<T> for Error {
    fn from(_: T) -> Self {
        Error
    }
}

fn af() -> impl core::future::Future<Output = Error> {
    async move { From::from(Error) }
}
/*
fn f() -> Error {
    From::from(Error)
}
*/
```

**Comment by nikomatsakis** [comment]
I've been investigating this. I'm not sure I understand yet what is going on, but I'm going to leave a few breadcrumbs as I go. To start, I think part of the problem lies on this line:

https://github.com/rust-lang/rust/blob/ed6468da160bd67a2ce0573427f09a98daff8c07/src/librustc_typeck/collect.rs#L1395

Specifically what is happening here is that we are executing the `type_of` query on a generator, and this is triggering us to do a full typeck. There is probably a good reason for this, but it's not generally how things are meant to work. This query is meant to be "modular", in that we can create the type for all items without type checking their bodies (this is possible because, for example, Rust requires us to fully specify the type signature for functions). I'm not quite sure why generators behave differently -- closures, for example, do not, as you can see if I give a bit more context:

https://github.com/rust-lang/rust/blob/ed6468da160bd67a2ce0573427f09a98daff8c07/src/librustc_typeck/collect.rs#L1393-L1400

Here, on lines 1398-1399, you can see that for a closure we create a type with "identity substitutions", which means a closure type like `Closure<P0...Pn>` where `Pi` is a reference to a type parameter. 

(I see that opaque types also behave differently; it seems like `tcx.type_of(opaque)` is used to fetch the concrete type of an opaque type. I actually think we should make that a distinct query, but I think it's neither here nor there.)

(I'm going to look a bit more at why generators behave this way now, not obvious to me that they must -- @Zoxc may have some details here?)

**Comment by nikomatsakis** [comment]
OK, I can confirm that the "naive fix" causes downstream errors, I haven't investigated those much yet:

```diff
@@ -1392,12 +1392,12 @@ fn type_of(tcx: TyCtxt<'_>, def_id: DefId) -> Ty<'_> {
         Node::Field(field) => icx.to_ty(&field.ty),

         Node::Expr(&hir::Expr { kind: hir::ExprKind::Closure(.., gen), .. }) => {
-            if gen.is_some() {
-                return tcx.typeck_tables_of(def_id).node_type(hir_id);
-            }
-
             let substs = InternalSubsts::identity_for_item(tcx, def_id);
-            tcx.mk_closure(def_id, substs)
+            if let Some(movability) = gen {
+                tcx.mk_generator(def_id, substs, movability)
+            } else {
+                tcx.mk_closure(def_id, substs)
+            }
         }

         Node::AnonConst(_) => {
```

I suspect they can be fixed though

**Comment by Zoxc** [comment]
@nikomatsakis Generator types contain the types of the values inside which are live across a yield statement, that is why `typeck_tables_of` is required there. We could move getting the interior type of the generator into a separate query, but that would only help if the problematic caller of `type_of` does not actually look at these.

**Comment by nikomatsakis** [comment]
@Zoxc I don't think that's sufficient reason by itself. Closure types also contain details that result from inference (e.g., the types that their upvars use), but `type_of` applied to a closure just gives you a fully generic type.

**Comment by tmandry** [comment]
We hit an instance of this in Fuchsia. Minimal example: ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=aee59c397f0d3659921c66d868cf1ff3))

```rust
mod queue {
    pub trait TryMerge {}
    
    impl TryMerge for () {}
    
    pub fn work_queue<W, C, WF>(work_fn: W)
    where
        W: Fn(C) -> WF,
        C: TryMerge,
    {}
}

impl queue::TryMerge for () {}

fn main() {
    queue::work_queue(move |_: ()| {
        async {}
    })
}
```

Gives 
```
error[E0283]: type annotations needed for `()`
```
instead of the expected error, which is what you get if you remove the async block:
```
error[E0119]: conflicting implementations of trait `queue::TryMerge` for type `()`
```
